src/app/qgisapp.cpp

```diff
@@ -10483,7 +10483,50 @@ void QgisApp::layerSubsetString()
 {
   QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( activeLayer() );
   if ( !vlayer )
+  {
+    // Try PG raster
+    QgsRasterLayer *rlayer = qobject_cast<QgsRasterLayer *>( activeLayer() );
+    if ( rlayer )
+    {
+      QgsRasterDataProvider *provider = rlayer->dataProvider();
+      if ( provider &&
+           provider->supportsSubsetString() )
+      {
+        // PG raster is the only one for now
+        if ( provider->name() == QStringLiteral( "postgresraster" ) )
+        {
+          // We need a vector for the sql editor
+          QgsDataSourceUri vectorUri { provider->dataSourceUri() };
```

Comment:
Well, this was of course uncharted territory, we have a raster layer that exposes some of the vector layer features.

So, here I had a few approaches:
1. ad a set of virtual vector methods in QgsRasterLayer and implement them in the "postgresraster" provider only, refactor query builder to take vectors **and** rasters (or even QgsMapLayer if the vector methods were moved to the base class)
2. make a vector layer out of a "postgresraster"  and feed the query builder, place the "hack" in App (where it won't fell lonesome)
3. implement a raster provider method that returns a (temporary) vector layer from the raster

Honestly, cost and time considerations aside, it seemed premature to go the hard way (1) only for a single provider, but I agree that it would be the natural refactoring one day.  I didn't like (2) either.

I think we should wait until we have another use case of a raster provider that supports a subset string before we go the hard way of abstracting that out.

Do you have anything in mind btw?